R/node-match.R

```diff
@@ -0,0 +1,454 @@
+#' Match patterns to a call
```

Comment:
I think that it is important to choose names that map as closely as possible to the teacher's mental model of grading. At first glance, `node_match()` doesn't seem to do that; it seems like something that might make more sense to a programmer than a teacher. On the other hand `node_match()` is a great name if we are educating users to think of each piece of student code as a network with nodes, for example to make it easier to navigate and isolate sections of the code. If this is the case, ignore my comment. I think it is fair to ask users to adopt new vocabulary so long as 1) there is a good reason for it, and 2) we teach the vocabulary and the bigger picture that it ties in to in a vignette.

If this was an arbitrary choice, I suggest something like code_match, call_match, function_match, pattern_match, etc.

Reply:
Agreed. `node_match()` as a name makes more sense within the landscape of low-level tools provided in rlang. It is still unclear the place `node_match()` will have in the final redpen UI, maybe we'll have a higher-level wrapper. If the teachers are going to use `node_match()`, we'll have to think about either providing a less technical-sounding synonym for it, or teach them about the tree structure of language calls as you suggest.

Reply:
FWIW I think teachers are going to have to internalise the tree model of code, at least to the level described in https://adv-r.hadley.nz/expressions.html